js/web/lib/wasm/jsep/webgpu/ops/gemm.ts

```diff
@@ -65,57 +63,75 @@ const createGemmProgramInfo = (inputs: readonly TensorView[], attributes: GemmAt
-    `@group(0) @binding(0) var<storage, read> a : array<${dataType}>;`,
-    `@group(0) @binding(1) var<storage, read> b : array<${dataType}>;`
+  const calculateAlpha = attributes.alpha === 1 ? '' : 'value *= uniforms.alpha;';
+  const [calculateCSnippet, broadcastM, broadcastN] =
+      inputs.length === 3 ? calculateC(M, N, inputs[2].dims) : ['', false, false];
+
+  const a = inputVariable('a', inputs[0].dataType, inputs[0].dims.length);
+  const b = inputVariable('b', inputs[1].dataType, inputs[1].dims.length);
+  const dataType = a.type.value;
+
+  const variables = [a, b];
+  const tensorDataType = tensorDataTypeEnumToString(inputs[0].dataType) as ProgramUniform['type'];
+  const programUniforms: ProgramUniform[] = [
+    {type: 'uint32', data: outputSize}, {type: 'uint32', data: M}, {type: 'uint32', data: N}, {type: 'uint32', data: K},
+    {type: tensorDataType, data: attributes.alpha}, {type: tensorDataType, data: attributes.beta},
+    ...createTensorShapeVariables(inputs[0].dims), ...createTensorShapeVariables(inputs[1].dims)
   ];
+
+  const inputDependencies: ProgramInputTensorInfoDependency[] = ['rank', 'rank'];
   if (inputs.length === 3) {
-    inputStorageBuffersDeclarations.push(`@group(0) @binding(2) var<storage, read> c : array<${dataType}>;`);
+    variables.push(inputVariable('c', inputs[2].dataType, inputs[2].dims.length));
+    programUniforms.push(...createTensorShapeVariables(inputs[2].dims));
+    inputDependencies.push('rank');
   }
-  const getShaderSource = (shaderHelper: ShaderHelper) => `
-  const M: u32 = ${M}u;
-  const N: u32 = ${N}u;
-  const K: u32 = ${K}u;
-  const alpha = ${dataType}(${attributes.alpha});
-  const beta = ${dataType}(${attributes.beta});
+  const output = outputVariable('output', inputs[0].dataType, outputShape.length);
+  variables.push(output);
+  programUniforms.push(...createTensorShapeVariables(outputShape));
 
-  ${inputStorageBuffersDeclarations.join('\n')}
-  @group(0) @binding(${inputs.length}) var<storage, read_write> output : array<${dataType}>;
+  const uniforms: UniformsArrayType = [
+    {name: 'outputSize', type: 'u32'}, {name: 'M', type: 'u32'}, {name: 'N', type: 'u32'}, {name: 'K', type: 'u32'},
+    {name: 'alpha', type: dataType as UniformDataElementType}, {name: 'beta', type: dataType as UniformDataElementType}
+  ];
+
+  const getShaderSource = (shaderHelper: ShaderHelper) => `
+  ${shaderHelper.registerUniforms(uniforms).declareVariables(...variables)}
 
   ${shaderHelper.mainStart()}
-    ${shaderHelper.guardAgainstOutOfBoundsWorkgroupSizes(outputSize)}
+    ${shaderHelper.guardAgainstOutOfBoundsWorkgroupSizes('uniforms.outputSize')}
 
-    let m = global_id.x / N;
-    let n = global_id.x % N;
+    let m = global_id.x / uniforms.N;
+    let n = global_id.x % uniforms.N;
 
     var value = ${dataType}(0);
-    for (var k: u32 = 0u; k<${K}u; k++) {
+    for (var k: u32 = 0u; k < uniforms.K; k++) {
       ${line}
     }
 
     ${calculateAlpha}
-    ${calculateC}
```

Comment:
For `calculateC`, can we use below snippet instead of the original `calculateC`?
```
let cOffset = ${c.broadcastedIndicesToOffset('uvec2(m, n)', output)};
value += beta * ${c.getByOffset('cOffset')};

Reply:
done